crates/sui-storage/src/fullnode_pending_tx_log.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! FullNodePendingTransactionLog is Write Ahead Log uesd in Transaction Orchestrator
+//! for transaction submission processing. It plays two roles:
+//! 1. At one time, a transaction is only processed once.
+//! 2. When Fullnode crashes and restarts, the pending transaction will be loaded and retried.
+
+use crate::mutex_table::MutexTable;
+use crate::write_ahead_log::{MUTEX_TABLE_SHARD_SIZE, MUTEX_TABLE_SIZE};
+use std::path::PathBuf;
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::{SuiError, SuiResult};
+use sui_types::messages::{TrustedTransction, VerifiedTransaction};
+use typed_store::traits::TypedStoreDebug;
+use typed_store::{rocks::DBMap, traits::Map};
+use typed_store_derive::DBMapUtils;
+
+pub type IsFirstRecord = bool;
+
+#[derive(DBMapUtils)]
+struct FullNodePendingTransactionTable {
+    logs: DBMap<TransactionDigest, TrustedTransction>,
+}
+
+pub struct FullNodePendingTransactionLog {
+    pending_transactions: FullNodePendingTransactionTable,
+    mutex_table: MutexTable<TransactionDigest>,
```

Comment:
TBH I think it would be simpler to just add pending_transactions to epoch tables, and have an in-memory mutex protected HashMap / BTreeMap of transactions to keep track of inflight transactions. It will make checking if a transaction is running easier. But I may be missing some requirements.

Reply:
This table shouldn't be per epoch, because FN needs to clear a transaction even if it knows about it in the last epoch. In the sunny days, this table should only have pending transactions in a short period of time.

> have an in-memory mutex protected HashMap / BTreeMap of transactions to keep track of inflight transactions.

This is what `MutexTable` is, no?

Reply:
I *guess* we are talking about the same thing [here](https://github.com/MystenLabs/sui/pull/6710#discussion_r1046375593)

Reply:
Ok, now I see what you mean, my previous comment was misguided - this retry log would be for transactions potentially in the pre-certificate formation stage, so the existing pending_certificates is not applicable.

However, in that case, I wonder why we need this at all? Shouldn't clients just retry if the fullnode crashes before returning an EffectsCert? Clients have to have retry code no matter what (they may fail due to network connectivity). What do we gain by adding statefulness here?

Reply:
resolved offline